core/utils/fork-tree/src/lib.rs

```diff
@@ -111,7 +111,16 @@ impl<H, N, V> ForkTree<H, N, V> where
 		)?;
 
 		if let Some(root) = new_root {
-			self.roots = vec![root.clone()];
+			let mut root = root.clone();
+
+			// we found the deepest ancestor of the finalized block, so we prune
+			// out any children that don't include the finalized block.
+			root.children.retain(|node| {
```

Comment:
there is only 1 child that can fulfill this precondition, right?

Reply:
Yes, do you think this should be more explicit?

Reply:
it is just going to be a little more inefficient

Reply:
```rust
let children = mem::replace(&mut root.children, Vec::new());
root.children = children.into_iter().filter(|node| ...).take(1).collect()
```